Update paper to R9 with suggested changes from LEWG

jesswong opened the pull request:
Initial PR with suggested changes from LEWG before forwarding to LWG

jesswong committed:
Update paper to R9 with suggested changes from LEWG

jesswong commented on asyncscope.md:
```diff
@@ -1268,7 +1332,7 @@ _`operation-state`_. The following algorithm determines which _Allocator_ to use
    - if `get_allocator(env)` is valid then `senv` is `env`;
    - otherwise, if `get_allocator(get_env(token.wrap(snd)))` is valid then `senv` is the expression
-     `@_JOIN-ENV_@(env, @_MAKE-ENV_@(get_allocator, alloc))`, where `alloc` is the chosen allocator;
+     `@_JOIN-ENV_@(env, prop(get_allocator, alloc))`, where `alloc` is the chosen allocator;
```
per Eric's comment: 

> oh, and MAKE-ENV should be prop now.

jesswong commented on asyncscope.md:
```diff
@@ -1369,18 +1433,18 @@ for the spawned sender to complete and then completes itself with the spawned se
      connect(
-         @_write-env_@(token.wrap(snd), senv),
-         @_spawn-future-receiver_@<@_completion-signatures-of_@<Sender>>{op}));
+         write_env(token.wrap(snd), senv),
```
per Eric's comment:

> also, write-env is probably becoming write_env this week.

jesswong commented on asyncscope.md:
```diff
@@ -1624,7 +1688,7 @@ sender auto join() noexcept;
 When `o` is started, the scope moves to either the open-and-joining or closed-and-joining state. `o` completes with
 `set_value()` when the scope moves to the joined state, which happens when the scope's count of outstanding operations
-drops to zero. `o` may complete synchronously if it happens to observe that the count of outstanding operations is
+drops to zero. `o` shall complete synchronously inline if it happens to observe that the count of outstanding operations is
```
per the poll results from our notes

jesswong commented on asyncscope.md:
```diff
@@ -2130,7 +2187,7 @@ the declaration of `run_loop`:
 >   template <async_scope_token Token, sender Sender>
 >     using @_wrapped-sender-from_@ =
->         decay_t<decltype(declval<Token&>().wrap(declval<Sender>()))>; // @_exposition-only_@
+>         remove_cvref_t<decltype(declval<Token&>().wrap(declval<Sender>()))>; // @_exposition-only_@
```
> Ruslan: I suggest you use remove_cvref_t instead of decay_t then.

jesswong commented on asyncscope.md:
```diff
@@ -2181,7 +2238,8 @@ method are not potentially throwing.
 [3]{.pnum} For a subexpression `snd`, let `Sndr` be `decltype((snd))` and let `sender<Sndr>` be true;
 `async_scope_token<Token>` is modeled only if, for an object, `token`, of type `Token`, the expression
-`token.wrap(snd)` is a valid expression and returns an object that satisfies `sender`.
+`token.wrap(snd)` is a valid expression and returns an object that satisfies `sender`, and does not add new
+ completion signatures.
```
> Lewis: The question is: What constraints do we want to put on implementations on wrap? We could say the the completion set must be a superset. It might add additional errors but should it be allowed to remove a completion signature from the input sender?

> Ian: I don’t have a good answer. This exists to allow the algorithm to rely on the fact that it can call wrap. If your implementation of wrap is surprising, you can’t use it with scope_token. I have to think about it more though. There should probably be a sentence that the output can’t add new completion signatures.

> Lewis: I have to think whether this is sufficient.


I wonder if this should be "and does not remove completion signatures"? Doesn't the counting_scope_token wrap introduce a set_stopped completion signature because of cancellation?

ispeters replied:
`counting_scope::token::wrap` doesn't add completions, it just injects a new stop token; if you nest, say, `just()` into a `counting_scope`, it doesn't become responsive to stop requests. The `set_stopped` that gets added is added by the fact that an attempt to nest may fail.

I think we want to say that `wrap` may not modify the input sender's completions.

jesswong commented on asyncscope.md:
```diff
@@ -2218,7 +2276,7 @@ struct @_nest-data_@ {
         : token(other.token) {
         if (token.try_associate()) {
-            sndr = other.sndr;
+            sndr.emplace(other.sndr);
```
> Robert: The assignment here should be sender.emplace.

jesswong commented on asyncscope.md:
```diff
@@ -2287,19 +2345,19 @@ following lambda:
         };
 
-        op_state(scope_token token, Rcvr& rcvr) noexcept
+        op_state(scope_token token, Rcvr& _rcvr) noexcept
```
> Lewis: There is a case where the member receiver is shadowing the parameter. It’s unclear which receiver you’re referring to. Rename one to make it more clear.

ispeters replied:
I'm unreasonably and irrationally bothered by the leading `_` so I'm going to just pick a different name. Sorry.

jesswong commented on asyncscope.md:
```diff
@@ -2381,20 +2429,35 @@ struct @_spawn-state-base_@ { // exposition-only
+    void set_value() && noexcept { state->@_complete_@(); }
+    void set_stopped() && noexcept { state->@_complete_@(); }
+
+    decltype(auto) get_env() const noexcept {
```
> Ian: spawn-receiver should respond with the environment computed above. That is missing.

Attempt to address this was modeled after https://eel.is/c++draft/exec#let-6

ispeters replied:
Turns out we don't need to do anything here:
* `get_env(rcvr)` returns `empty_env{}` if `rcvr.get_env()` is invalid; and
* `empty_env{}` is fine for this receiver because we use `write_env` below (in `makeSender`) to provide the appropriate environment to the spawned operation.

We do, however, need to provide `using receiver_concept = receiver_t;`.

jesswong commented on asyncscope.md:
```diff
@@ -2415,6 +2478,7 @@ private:
     @_alloc-t_@ alloc;
     @_op-t_@ op;
     Token token;
+    bool associated(false);
```
> Ian: There’s a bug here. In run(), I can’t delegate complete(). If we fail to associate we want to perform only what is inside the braces but we don’t want to disassociate.

Similar implementation to nest

ispeters replied:
I'm thinking we can save storing the bool if we just factor "destroy and deallocate" out of "complete".

jesswong commented on asyncscope.md:
```diff
@@ -2381,20 +2429,35 @@ struct @_spawn-state-base_@ { // exposition-only
     @_spawn-state-base_@* state; // exposition-only
-    void set_value() && noexcept { state->complete(); }
-    void set_stopped() && noexcept { state->complete(); }
+    void set_value() && noexcept { state->@_complete_@(); }
+    void set_stopped() && noexcept { state->@_complete_@(); }
```
> Ruslan: state-erorr-complete is exposition only? It should be italicized.

jesswong commented on asyncscope.md:
````diff
@@ -2447,21 +2511,24 @@ private:
 
 ```cpp
-    auto token = this->token;
+    auto token = std::move(this->token);
````
> Robert: Why does it copy the token in complete() instead of move?

jesswong commented on asyncscope.md:
```diff
@@ -2505,51 +2572,39 @@ either the result of the eagerly-started input sender or with `set_stopped` if t
 
-- if `get_stop_token(env)` is well defined then `stok` receives stop requests sent from the returned future and any stop
-  requests sent to the stop token returned from `get_stop_token(env)`;
+- `stok` receives stop requests sent from the returned future and any stop requests sent to the stop token returned from `get_stop_token(env)`;
```
> Robert: In what situation is get_stop_token(env) not well defined? You don’t need the second bulletpoint after if well defined.

jesswong commented on asyncscope.md:
```diff
@@ -2505,51 +2572,39 @@ either the result of the eagerly-started input sender or with `set_stopped` if t
 template <class Sigs>
 struct @_spawn-future-state-base_@ { // @_exposition-only_@
-    variant</* see below */> result; // @_exposition-only_@
+    std::variant</* see below */> result; // @_exposition-only_@
```
> Ruslan: There should be a std:: in front of variant. It is not exposition-only

jesswong commented on asyncscope.md:
```diff
@@ -2505,51 +2572,39 @@ either the result of the eagerly-started input sender or with `set_stopped` if t
 - `monostate` as the first element;
 - for each completion signature in `Sigs` with a completion tag `cpo_t` and parameter types `P...` an element of type
-  `tuple<cpo_t, remove_cvref_t<P>...>`; and
+  `@_decayed-tuple_@<cpo_t, Ps...>`; and
```
per Eric's comment:

> and replace tuple<cpo, remove_cvref_t<Ts>...> with decayed-tuple<cpo, Ts...> generally (decayed-tuple in italics)

jesswong commented on asyncscope.md:
```diff
@@ -2561,18 +2616,20 @@ struct @_spawn-future-receiver_@ { // @_exposition-only_@
         }
         catch (...) {
-            state->result.emplace<tuple<set_error_t, exception_ptr>>(set_error, std::current_exception());
+            if constexpr () { // TODO: how to express to not throw if things are non throwable
```
> Lewis: I think existing wording uses if-constexpr.

Example: https://eel.is/c++draft/exec#schedule.from-11

jesswong commented on asyncscope.md:
```diff
@@ -2652,9 +2725,9 @@ private:
+  - then `this->@_destroy_@()` is invoked in the _op-t_ destructor.
 - otherwise, an invocation of _`abandon`_ happened-before this invocation of _`complete`_ and `this->@_destroy_@()`
-  is invoked.
+  is invoked in the _op-t_ destructor.
```
> ACTION: In section 12, modify the wording to call “destroy” in the op_state dtor

ispeters replied:
I think you misunderstood which operation state needs to do this in its destructor; I think I've addressed this now.

jesswong committed:
Address some missing similar changes

jesswong committed:
remove empty space

jesswong commented on asyncscope.md:
```diff
@@ -2505,51 +2571,36 @@ either the result of the eagerly-started input sender or with `set_stopped` if t
-- if `get_stop_token(env)` is well defined then `stok` receives stop requests sent from the returned future and any stop
-  requests sent to the stop token returned from `get_stop_token(env)`;
-- otherwise `stok` only receives stop requests sent from the returned future.
+requests sent from the returned future and any stop requests sent to the stop token returned from `get_stop_token(env)`
```
> Robert: In what situation is get_stop_token(env) not well defined? You don’t need the second bulletpoint after if well defined.

jesswong committed:
more nits

jesswong committed:
nits

jesswong commented:
After speaking to Eric, we won't be adding the completion signatures for spawn/spawn_future/nest. Per Eric:

> @Ian Petersen none of the sender algorithms overtly specify their completion signatures. there is generic wording about the completion signatures corresponding to the potentially evaluated completion operations. you don't need to say anything else.

jesswong commented on asyncscope.md:
```diff
@@ -1624,7 +1687,7 @@ sender auto join() noexcept;
 When `o` is started, the scope moves to either the open-and-joining or closed-and-joining state. `o` completes with
 `set_value()` when the scope moves to the joined state, which happens when the scope's count of outstanding operations
-drops to zero. `o` may complete synchronously if it happens to observe that the count of outstanding operations is
+drops to zero. `o` will complete synchronously inline if it happens to observe that the count of outstanding operations is
```
per the poll results from our notes

jesswong commented on asyncscope.md:
```diff
@@ -2505,51 +2571,36 @@ either the result of the eagerly-started input sender or with `set_stopped` if t
 - `monostate` as the first element;
 - for each completion signature in `Sigs` with a completion tag `cpo_t` and parameter types `P...` an element of type
-  `tuple<cpo_t, remove_cvref_t<P>...>`; and
+  `@_decayed-tuple_@<cpo_t, P...>`; and
```
per Eric's comment:

> and replace tuple<cpo, remove_cvref_t...> with decayed-tuple<cpo, Ts...> generally (decayed-tuple in italics)

jesswong committed:
Change decayed_tuple to decayed-tuple

ispeters committed:
Clarify wrap() can't modify completions

ispeters committed:
Add missing punctuation

Add a missing `*` when completing a receiver with `set_stopped` and
terminate a bullet point with a period.

ispeters committed:
Rename _rcvr to r

ispeters committed:
Fixup spawn-receiver; add receiver_concept, remove get_env

ispeters committed:
Avoid storing an "associated" bool in spawn-state

Factor the "destroy and deallocate" logic out of spawn-state::complete
into a new expo-only method named destroy, and then call it from the
else branch in run when token.try_associate() fails.

ispeters committed:
Fix up the spawn-future-receiver

* add the receiver_concept type alias
* ensure all the "tuples" are _decayed-tuple_
* make the try-catch blocks conditional on whether they're necessary

ispeters committed:
Fixups to spawn_future

One more tuple -> decayed-tuple change, and address the fact that the
invocation of `visit` on the result variant was wrong: the argument to
the callable will be a decayed-tuple that needs to be decomposed into
its parts with std::apply (unless it's monostate...).

ispeters committed:
Make sure we JOIN-ENV in the right order

JOIN-ENV(a, b) produces an environment that queries a before b so we
need to make sure we're constructing our joined environments in the
right order.

ispeters committed:
JOIN-ENV appears to be a binary function

We were doing JOIN-ENV(a, b, c), but JOIN-ENV is binary so update to
JOIN-ENV(JOIN-ENV(a, b), c).

ispeters committed:
Update section 5.5's description of spawn_future's stop_token

A previous commit updated the wording for how `spawn_future` computes a
stop token; this commit updates the corresponding prose in section 5.5.

ispeters committed:
Delete the spawned opstate in the future op's dtor

Ensure that it's the future-sender's opstate that destroys the spawned
operation (via a call to abandon) when the future consumes the result of
the spawned work.

ispeters committed:
Incorporate LWG feedback

A whole whack of changes from Wednesday and Friday sessions in LWG in
Hagenberg, Austria.

jesswong commented on asyncscope.md:
```diff
@@ -2345,128 +2396,246 @@ object.
+
+    @_spawn-future-state_@(Alloc alloc, Sender&& sndr, Token token) // @_exposition only_@
+        : @_alloc_@(std::move(alloc)),
+          @_
```
is this extra?

ispeters replied:
Looks like I got distracted; I'll see if I can reconstruct what was supposed to be here.

jesswong commented on asyncscope.md:
````diff
@@ -2619,110 +2751,61 @@ private:
 ```cpp
     if (token.try_associate()) {
-        op.start();
+        op.start()
````
```suggestion
        op.start();
```

ispeters committed:
Final tweaks

* Address @jesswong's PR feedback
* Reformat a few over-long lines
* Apply some LWG advice like "no braces around single-line if bodies"

ispeters committed:
Update to a P-number
